src/sentry/similarity/index.py

```diff
@@ -28,42 +27,58 @@ def __init__(self, cluster, namespace, signature_builder, bands, interval, reten
         self.interval = interval
         self.retention = retention
 
-    def __build_signatures(self, items):
-        data = defaultdict(
-            lambda: [Counter() for _ in xrange(self.bands)],
-        )
-
-        for idx, features in items:
-            bands = map(
-                ','.join, band(
-                    self.bands,
-                    map(
-                        '{}'.format,
-                        self.signature_builder(features),
-                    ),
-                )
-            )
-
-            for i, bucket in enumerate(bands):
-                data[idx][i][bucket] += 1
-
-        arguments = [len(data)]
-        for idx, bands in data.items():
-            arguments.append(idx)
-            for buckets in bands:
-                arguments.append(len(buckets))
-                for bucket, count in buckets.items():
-                    arguments.extend([
-                        bucket,
-                        count,
-                    ])
+    def _build_signature_arguments(self, features):
+        if not features:
+            return [0] * self.bands
 
+        arguments = []
+        for bucket in band(self.bands, self.signature_builder(features)):
+            arguments.extend([1, ','.join(map('{}'.format, bucket)), 1])
         return arguments
 
     def _get_connection(self, scope):
         return self.cluster.get_local_client_for_key(scope)
 
-    def classify(self, scope, items, timestamp=None):
+    def _as_search_result(self, indices, result):
+        replacements = {
+            -1: None,
+            -2: 0,
+        }
+
+        def get_comparison_key((key, scores)):
+            scores = filter(
+                lambda score: score is not None,
+                map(
+                    lambda score: replacements.get(score, score),
+                    scores,
+                ),
+            )
+
+            return (
+                sum(scores) / len(scores) * -1,  # average score
+                len(scores) * -1,  # number of indexes with scores, tiebreaker
+                key,  # lex sort, tiebreaker
+            )
+
+        # TODO: Clean this up, don't map so much.
```

Comment:
There has to be a better way to do this but I haven't put much thought into it yet

Reply:
I think honestly this could just do with some variable assignments. I actually want to have a discussion about this more broadly, since I think anytime you're mapping with more than two levels of additional func calls it becomes super confusing, and we tend to do that a good bit.

This could be

```py
def score_transformer(key, score):
  return (key, map(float, scores))

def score_reducer(key, scores):
  return (key, map(lambda score: score if score >= 0 else None, scores))

results = map(score_transformer, result)
results = sort(results, key=get_comparison_key)

return map(score_reducer, results)
```

This generally reads better to me

Reply:
Yeah, I'm generally unhappy with this entire function (other than the fact that it works.)

Will improve

Reply:
Didn't exactly do the changes you suggested (to be honest, that to snippet me is more of a lateral move because it doesn't really explain what/why the functions exist) but updated this to be more coherent in 9e3f869345019f292db4ea3eee55487c2cca3a3e. I still don't love it but it's significantly better now.